File: kotlinapp/app/src/main/java/com/emusic/app/ui/components/Skeleton.kt
package com.emusic.app.ui.components

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.composed
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.unit.dp

/** Efecto shimmer (barrido de brillo) para placeholders de carga. */
fun Modifier.shimmer(): Modifier = composed {
    val transition = rememberInfiniteTransition(label = "shimmer")
    val x by transition.animateFloat(
        initialValue = 0f,
        targetValue = 1200f,
        animationSpec = infiniteRepeatable(
            animation = tween(1100, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "x"
    )
    // Semi-transparente: los placeholders se ven suaves, dejando entrever el fondo.
    val base = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.12f)
    val highlight = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.28f)
    background(
        brush = Brush.linearGradient(
            colors = listOf(base, highlight, base),
            start = Offset(x - 350f, 0f),
            end = Offset(x, 0f)
        )
    )
}

private fun Modifier.skeletonBox(shape: androidx.compose.ui.graphics.Shape) =
    this.clip(shape).shimmer()

/** Fila placeholder con forma de TrackItem (carátula + título + artista). */
@Composable
fun TrackRowSkeleton(modifier: Modifier = Modifier) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Box(Modifier.size(56.dp).skeletonBox(RoundedCornerShape(6.dp)))
        Spacer(Modifier.width(12.dp))
        Column(Modifier.weight(1f)) {
            Box(Modifier.fillMaxWidth(0.7f).height(14.dp).skeletonBox(RoundedCornerShape(4.dp)))
            Spacer(Modifier.height(8.dp))
            Box(Modifier.fillMaxWidth(0.4f).height(12.dp).skeletonBox(RoundedCornerShape(4.dp)))
        }
    }
}

/** Lista de filas placeholder para mientras se cargan resultados. */
@Composable
fun TrackListSkeleton(count: Int = 8, modifier: Modifier = Modifier) {
    Column(modifier = modifier.fillMaxWidth()) {
        repeat(count) { TrackRowSkeleton() }
    }
}

File: kotlinapp/iosApp/iosApp/Skeleton.swift
import SwiftUI

/// Puerto de `ui/components/Skeleton.kt`: placeholders con barrido de brillo para
/// mientras cargan las listas. Se usan en Inicio, Buscar y Biblioteca, igual que en
/// Android.

/// Barrido de brillo infinito. En Compose es `rememberInfiniteTransition` con
/// `tween(1100, LinearEasing)` en bucle; aca es una animacion repetida sobre el
/// desplazamiento del gradiente.
private struct ShimmerModifier: ViewModifier {

    @State private var phase: CGFloat = -1

    /// Semitransparentes, para que los placeholders dejen entrever el fondo.
    private let base = EMusicColor.onSurfaceVariant.opacity(0.12)
    private let highlight = EMusicColor.onSurfaceVariant.opacity(0.28)

    func body(content: Content) -> some View {
        content
            .background(
                GeometryReader { geo in
                    LinearGradient(
                        colors: [base, highlight, base],
                        startPoint: .leading,
                        endPoint: .trailing
                    )
                    .frame(width: geo.size.width * 3)
                    .offset(x: phase * geo.size.width * 1.5)
                }
            )
            .onAppear {
                withAnimation(.linear(duration: 1.1).repeatForever(autoreverses: false)) {
                    phase = 1
                }
            }
    }
}

extension View {
    func shimmer() -> some View {
        modifier(ShimmerModifier())
    }
}

/// Caja con forma redondeada y shimmer, la pieza basica de los placeholders.
private struct SkeletonBox: View {
    var width: CGFloat?
    var height: CGFloat
    var radius: CGFloat

    var body: some View {
        RoundedRectangle(cornerRadius: radius)
            .fill(Color.clear)
            .frame(width: width, height: height)
            .shimmer()
            .clipShape(RoundedRectangle(cornerRadius: radius))
    }
}

/// Fila placeholder con la forma de `TrackItem`: caratula 56, titulo al 70% del ancho
/// y artista al 40%. Mismas medidas que `TrackRowSkeleton` en Compose.
struct TrackRowSkeleton: View {
    var body: some View {
        HStack(spacing: 12) {
            SkeletonBox(width: 56, height: 56, radius: 6)

            GeometryReader { geo in
                VStack(alignment: .leading, spacing: 8) {
                    SkeletonBox(width: geo.size.width * 0.7, height: 14, radius: 4)
                    SkeletonBox(width: geo.size.width * 0.4, height: 12, radius: 4)
                }
                .frame(height: geo.size.height, alignment: .center)
            }
            .frame(height: 34)
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 8)
    }
}

/// Lista de filas placeholder. Android usa 8 por defecto y 5 en Inicio.
struct TrackListSkeleton: View {
    var count: Int = 8

    var body: some View {
        VStack(spacing: 0) {
            ForEach(0..<count, id: \.self) { _ in
                TrackRowSkeleton()
            }
        }
    }
}
